### PR TITLE
feat(api): Get Customise page data

### LINKED ISSUE #2519: Customize
All the data inside the page with GET & SET

### PR DESCRIPTION
**Pull request by soham4abc**
## Description

This PR introduces an endpoint to fetch data for the customize page from admin panel 

THIS PR IS WIP and to be updated according to the requirements.

## How to test

send a `GET` request to `localhost/repo/api/v1/customise`

## Screenshots
Screenshot from 2023-07-04 20-50-06

closes #2519

<a href="https://gitpod.io/#https://github.com/fossology/fossology/pull/2520"></a>

### CODE REVIEW

**Comment by soham4abc** [comment]
image
Updated response for `/customise` endpoint.

Please do have a look @GMishx

**Comment by soham4abc** [comment]
This PR also updates the Database schema to update description of `vartypes` in `sysconfig` table
@GMishx

File: src/www/ui/api/Controllers/CustomiseController.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    if (Auth::isAdmin()) {
+      $finalVal = [];
+      $returnVal = $this->configContoller->getConfigData();
+      $typeArr = array(
+        '1' => "int",
+        '2' => "text",
+        '3' => 'textarea',
+        '4' => 'password',
+        '5' => 'dropdown',
+        '6' => 'boolean'
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CustomiseController.php:66]: Please use consistent quote types.

File: src/www/ui/api/Controllers/CustomiseController.php [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        $finalVal[$row['variablename']] = array(
+          "value" => $value,
+          "type" => $type,
+          "label" => $row['ui_label'],
+          "group_name" => $row['group_name'],
+          "group_order" => intval($row['group_order']),
+          "option_value" => $row["option_value"]
+        );
+      }
+      return ($response->withJson($finalVal, 200));
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CustomiseController.php:84]:
```suggestion
      return $response->withJson($finalVal, 200);
```

File: src/www/ui/api/documentation/openapi.yaml [hunk 3]
```diff
@@ -3434,6 +3459,33 @@ components:
...
+        key:
+          description: field key
+          type: object
+          properties:
+            value:
+              description: field value
+              type: string
+            type:
+              description: field type
+              type: string
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:3475]: Please document different types as enum

File: src/www/ui/api/documentation/openapi.yaml [hunk 4]
```diff
@@ -3434,6 +3459,33 @@ components:
...
+            group_name:
+              description: group name for the field
+              type: string
+            group_order:
+              description: group order of the field
+              type: integer
+            option_value:
+              description: If type is 'dropdown', provide options in format op1{val1}|op2{val2}|...
+              type: integer
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:3487]:
```suggestion
              type: string
```

File: src/www/ui/core-schema.dat [hunk 5]
```diff
@@ -1805,7 +1805,7 @@
   $Schema["TABLE"]["sysconfig"]["ui_label"]["ADD"] = "ALTER TABLE \"sysconfig\" ADD COLUMN \"ui_label\" varchar(60)";
   $Schema["TABLE"]["sysconfig"]["ui_label"]["ALTER"] = "ALTER TABLE \"sysconfig\" ALTER COLUMN \"ui_label\" SET NOT NULL";
 
-  $Schema["TABLE"]["sysconfig"]["vartype"]["DESC"] = "COMMENT ON COLUMN \"sysconfig\".\"vartype\" IS 'variable type.  1=int, 2=text, 3=textarea, 4=password, 5=dropdown'";
+  $Schema["TABLE"]["sysconfig"]["vartype"]["DESC"] = "COMMENT ON COLUMN \"sysconfig\".\"vartype\" IS 'variable type.  1=int, 2=text, 3=textarea, 4=password, 5=dropdown 6=boolean'";
```
Comment (GMishx) [inline src/www/ui/core-schema.dat:1808]:
```suggestion
  $Schema["TABLE"]["sysconfig"]["vartype"]["DESC"] = "COMMENT ON COLUMN \"sysconfig\".\"vartype\" IS 'variable type.  1=int, 2=text, 3=textarea, 4=password, 5=dropdown, 6=boolean'";
```

**Review by GMishx — CHANGES_REQUESTED** [review]

File: src/www/ui/admin-config.php [hunk 6]
```diff
@@ -32,6 +32,17 @@ function __construct()
     $this->dbManager = $GLOBALS['container']->get('db.manager');
   }
 
+  /**
+   * \brief Fetch configuration rows.
+   */
+  function getConfigData()
```
[the remark this answers is not in the document]
Comment (soham4abc) [reply src/www/ui/admin-config.php:38]: So, should I be creating a separate DAO for the admin functions named AdminDAO? and create a method there itself to get the config data in the desired output format?

**Comment by soham4abc** [comment]
PR updated as per meeting discussions on 12/07/2023

@GMishx Please do have a look

File: src/lib/php/Dao/SysConfigDao.php [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+  /**
+   * \brief Fetch configuration rows.
+   */
+  function getConfigData()
+  {
+    global $PG_CONN;
+    $sql = "select * from sysconfig order by group_name, group_order";
+    $result = pg_query($PG_CONN, $sql);
+    return (pg_fetch_all($result));
```
Comment (GMishx) [inline src/lib/php/Dao/SysConfigDao.php:42]: Use modern DBManager.
```suggestion
    $sql = "SELECT * FROM sysconfig ORDER BY group_name, group_order";
    return $this->dbManager->getRows($sql);
```

File: src/lib/php/Dao/SysConfigDao.php [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  /**
+   * Get all customise information for admin
+   * @return array
+   */
+  public function getCustomiseData($data)
```
Comment (GMishx) [inline src/lib/php/Dao/SysConfigDao.php:49]: Please document in docstring what `$data` should look like.

File: src/lib/php/Dao/SysConfigDao.php [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  {
+    $finalVal = [];
+    foreach ($data as $row) {
+      $type = self::TYPE_MAP[$row['vartype']];
+      if ($type == "int") {
+        $value = intval($row['conf_value']);
+      } else {
```
Comment (GMishx) [inline src/lib/php/Dao/SysConfigDao.php:56]: Keeping the return type consistent as string will help consumer of the endpoint by using single class to read the data.

File: src/www/ui/api/documentation/openapi.yaml [hunk 10]
```diff
@@ -2491,7 +2491,32 @@ paths:
                 $ref: '#/components/schemas/Info'
         default:
           $ref: '#/components/responses/defaultResponse'
-
+
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:2494]:
```suggestion

```

File: src/www/ui/api/documentation/openapi.yaml [hunk 11]
```diff
@@ -3981,6 +4006,41 @@ components:
...
+            group_name:
+              description: group name for the field
+              type: string
+            group_order:
+              description: group order of the field
+              type: integer
+            option_value:
+              description: If type is 'dropdown', provide options in format op1{val1}|op2{val2}|...
+              type: string
+
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:4043]: This structure no longer matches the current response. Please update.

**Review by GMishx — CHANGES_REQUESTED** [review]

**Comment by soham4abc** [comment]
Made the requested changes. Please do have a look @GMishx 
Thank you!

File: src/www/ui/api/documentation/openapi.yaml [hunk 12]
```diff
@@ -2492,6 +2492,31 @@ paths:
...
           $ref: '#/components/responses/defaultResponse'
 
+  /customise:
+    get:
+      operationId: getCustomiseData
+      tags:
+        - Admin
+      summary: Get the Admin configureation data
+      description: >
+        Returns the admin configureation data from the server
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:2502]:
```suggestion
      summary: Get the Admin configuration data
      description: >
        Returns the admin configuration data from the server
```

File: src/www/ui/api/documentation/openapi.yaml [hunk 13]
```diff
@@ -3981,6 +4006,40 @@ components:
         isCandidate:
           description: Is the license a candidate?
           type: boolean
+    Customise:
+      description: Admin customise ddata object
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:4010]:
```suggestion
      description: Admin customise data object
```

File: src/www/ui/api/documentation/openapi.yaml [hunk 14]
```diff
@@ -3981,6 +4006,40 @@ components:
...
+        group_name:
+          description: group name for the field
+          type: string
+        group_order:
+          description: group order of the field
+          type: integer
+        option_value:
+              description: If type is 'dropdown', provide options in format op1{val1}|op2{val2}|...
+              type: string
+
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:4042]: Small formatting changes needed.
```suggestion
          - boolean
        label:
          description: field label
          type: string
        group_name:
          description: group name for the field
          type: string
        group_order:
          description: group order of the field
          type: integer
        option_value:
          description: If type is 'dropdown', provide options in format op1{val1}|op2{val2}|...
          type: string

```

**Review by GMishx — CHANGES_REQUESTED** [review]

**Review by GMishx — APPROVED** [review]
Changes looks good.

**Pull request merged**